Docs for the 0.22 release (#2307)
kestra-io/docs: content/docs/06.enterprise/02.governance/read-only-secrets.md

---
title: Read-only Secrets
icon: /docs/icons/kestra.svg
editions: ["EE"]
---

Integrate external secrets managers in a read-only mode.

## Read-only secrets

When integrating an external [secrets manager](secrets-manager.md) with Kestra, you may want to ensure that those secrets cannot be modified within Kestra, maintaining immutability.
Currently, read-only secrets can be configured for [AWS Secret Manager](secrets-manager.md#aws-secret-manager-configuration), [Azure Key Vault](secrets-manager.md#azure-key-vault-configuration), [Google Secret Manager](secrets-manager.md#google-secret-manager-configuration), and [Vault](secrets-manager.md#vault-configuration).

## Configure read-only secrets

> Read-only secrets can be configured globally in the configuration file as well as enabled from the UI in the [Namespace](../../04.workflow-components/02.namespace.md) edit page. To turn on for a specific Namespace, toggle on in the UI.

Reviewer: ```suggestion
Read-only secrets can be configured globally in the configuration file as well as enabled from the UI in the [Namespace](../../04.workflow-components/02.namespace.md) edit page and Create [Tenant](./tenants.md) page. To turn on for a specific Namespace, toggle on in the UI.
```

![read-only-secrets-1](/docs/enterprise/read-only-secrets-1.png)

Secrets will now have a lock icon to show that they cannot be edited from Kestra, and the **Create New Secret** button in the top right corner that would otherwise be present is unavailable.

![read-only-secrets-4](/docs/enterprise/read-only-secrets-4.png)

To configure globally, add `readOnly: true` to the configuration of your external secret manager like in the examples below.

### AWS Secret Manager
 
The following example shows the configuration for AWS Secret Manager with read-only:

```yaml
kestra:
  secret:
    type: aws-secret-manager
    readOnly: true
    awsSecretManager:
      accessKeyId: mysuperaccesskey
      secretKeyId: mysupersecretkey

>       sessionToken: mysupersessiontoken

Reviewer: Optional
```suggestion
```

        Paste here the contents of the service account JSON key file
```

### Vault

With [Vault](./secrets-manager.md#vault-configuration), secrets are stored in a unique structure that can vary depending on the organization and version of Vault. Typically, there is a Secret Engine that hosts different Secrets with specific paths. Those Secrets are the paths to subkeys that are the actual key value pairs such as Username or Password to a service (e.g., `MY_SECRET = MY_SECRET_PASSWORD`). The following demonstrates a visual representation of this structure:

```
secret/
  ├── app1/
  │   ├── db/ <-- SECRET
  │   │   ├── DATBASE_USERNAME    # Subkey
  │   │   ├── DATABASE_PASSWORD   # Subkey
  │   ├── api/ <-- SECRET
  │       ├── keys                # Subkey
  │       ├── API_TOKEN           # Subkey
  ├── app2/
      ├── config
```

- `secret`: This is the secret engine.
- `app1` and `app2`: These are the path names to the secrets. This could be for example separate business units or applications.
- `db`, `api`, and `config`: These are the secret names visible in the Kestra UI. `api` could be the Vault Secret that contains all API Keys for an application's external services.

> - `DATBASE_USERNAME`, `DATABASE_PASSWORD`, `keys`, `API_TOKEN`: These are the `subkey` key value pairs that can be used in a Kestra flow.

Reviewer: ```suggestion
- `DATABASE_USERNAME`, `DATABASE_PASSWORD`, `keys`, `API_TOKEN`: These are the `subkey` key value pairs that can be used in a Kestra flow.
```

> - `DATBASE_USERNAME`, `DATABASE_PASSWORD`, `keys`, `API_TOKEN`: These are the `subkey` key value pairs that can be used in a Kestra flow.

Reviewer: ```suggestion
- `DATABASE_USERNAME`, `DATABASE_PASSWORD`, `keys`, `API_TOKEN`: These are the `subkey` key value pairs that can be used in a Kestra flow.
```

> With the above example structure, if we only need secrets for `app1`, our configuration in Kestra looks as follows with the added property `secretPathPrefix`:

Reviewer: Maybe a note about how the prefix is not going to be used in the global config? probably not in the tenant one but check

---
title: Read-only Secrets
icon: /docs/icons/kestra.svg
editions: ["EE"]
---

Integrate external secrets managers in a read-only mode.

> 

Reviewer: Read Only Secrets video

```suggestion

<div class="video-container">
    <iframe src="https://www.youtube.com/embed/CJYoGfWDIO8?si=b4z3MZaHf_uddM_0" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>
</div>

```

---
title: Read-only Secrets
icon: /docs/icons/kestra.svg
editions: ["EE"]

> ---

Reviewer: ```suggestion
version: "0.22.0"
---
```

Integrate external secrets managers in a read-only mode.

<div class="video-container">

>     <iframe src="https://www.youtube.com/embed/CJYoGfWDIO8?si=b4z3MZaHf_uddM_0" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

Reviewer: ```suggestion
    <iframe src="https://www.youtube.com/embed/uxFyE1nsMlU?si=X3nUxXwfAu4jCElc" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>
```

Reviewer: Fixed removing a bug that's been solved now

![read-only-secrets-8](/docs/enterprise/read-only-secrets-8.png)

To turn on for a specific Namespace, toggle the setting on in the **Dedicated secrets manager** configuration of the **Edit** tab.

![read-only-secrets-1](/docs/enterprise/read-only-secrets-1.png)

Secrets will now have a lock icon to show that they cannot be edited from Kestra, and the **Create New Secret** button in the top right corner that would otherwise be present is unavailable.

![read-only-secrets-4](/docs/enterprise/read-only-secrets-4.png)

To configure globally, add `readOnly: true` to the configuration of your external secret manager like in the examples below.

### AWS Secret Manager
 
For compatibility with Kestra, ensure that your AWS screts are stored as plain text in AWS Secrets Manager and not as key value pairs. The following example shows the configuration for AWS Secret Manager with read-only:

```yaml
kestra:
  secret:
    type: aws-secret-manager
    readOnly: true
    awsSecretManager:
      accessKeyId: mysuperaccesskey
      secretKeyId: mysupersecretkey
      region: us-east-1
```

### Azure Key Vault

> The following example shows the configuration for Azure Key Vault with read-only:

Reviewer 2: ```suggestion
The following example shows the configuration for Azure Key Vault with a read-only secrets backend:
```

```yaml
kestra:
  secret:
    type: azure-key-vault
    readOnly: true
    azureKeyVault:
      clientSecret:
        tenantId: "id"
        clientId: "id"
        clientSecret: "secret"
```

### Google Secret Manager

> The following example shows the configuration for Google Secret Manager with read-only:

Reviewer 2: ```suggestion
The following example shows the configuration for Google Secret Manager with a read-only secrets backend:
```

With the above example structure, if we only need secrets for `app1`, our configuration in Kestra looks as follows with the added property `secretPathPrefix`:

```yaml
address: https://my-vault:8200/
rootEngine: secret
secretPathPrefix: app1
token:
  token: my-vault-access-token
```

This configuration gives Kestra access to the `db` and `api` secrets as they are the secrets on the `app1` path. In a flow, to access the value for the subkey `API_TOKEN`, you write the `secret()` function with the specified parameters `{{ secret('api', subkey='API_TOKEN') }}`.

## Vault full example

The following steps walk through a full example configuring Vault as your secret manager with read-only secrets enabled. This example uses [KV Secrets Engine - Version 2 with Vault Enterprise](./secrets-manager.md#kv-secrets-engine---version-2) so `rootEngine` and `namespace` are used as optional properties.

In Vault, we have a Secrets Engine named `business-unit` in the `admin` namespace that hosts the path to our database password that we want to use to [add a table and populate with data in Neon](../../15.how-to-guides/neon.md).

![read-only-secrets-2](/docs/enterprise/read-only-secrets-2.png)

> In Kestra, we can now navigate to the Namespace we want to set up Vault as a secrets manager and enter the configuration details

Reviewer 2: ```suggestion
In Kestra, we can now navigate to the Namespace we want to set up Vault as a secrets manager for and enter the configuration details:
```

  │       ├── keys                # Subkey
  │       ├── API_TOKEN           # Subkey
  ├── app2/
      ├── config
```

- `secret`: This is the secret engine.
- `app1` and `app2`: These are the path names to the secrets. This could be for example separate business units or applications.
- `db`, `api`, and `config`: These are the secret names visible in the Kestra UI. `api` could be the Vault Secret that contains all API Keys for an application's external services.
- `DATABASE_USERNAME`, `DATABASE_PASSWORD`, `keys`, `API_TOKEN`: These are the `subkey` key value pairs that can be used in a Kestra flow.

With the above example structure, if we only need secrets for `app1`, our configuration in Kestra looks as follows with the added property `secretPathPrefix`:

```yaml
address: https://my-vault:8200/
rootEngine: secret
secretPathPrefix: app1
token:
  token: my-vault-access-token
```

This configuration gives Kestra access to the `db` and `api` secrets as they are the secrets on the `app1` path. In a flow, to access the value for the subkey `API_TOKEN`, you write the `secret()` function with the specified parameters `{{ secret('api', subkey='API_TOKEN') }}`.

## Vault full example

> The following steps walk through a full example configuring Vault as your secret manager with read-only secrets enabled. This example uses [KV Secrets Engine - Version 2 with Vault Enterprise](./secrets-manager.md#kv-secrets-engine---version-2) so `rootEngine` and `namespace` are used as optional properties.

Author: ```suggestion
The following steps are a full example of configuring Vault as your secret manager with read-only secrets enabled. This example uses [KV Secrets Engine - Version 2 with Vault Enterprise](./secrets-manager.md#kv-secrets-engine---version-2), so `rootEngine` and `namespace` are used as optional properties.
```

In Vault, we have a Secrets Engine named `business-unit` in the `admin` namespace that hosts the path to our database password that we want to use to [add a table and populate with data in Neon](../../15.how-to-guides/neon.md).

![read-only-secrets-2](/docs/enterprise/read-only-secrets-2.png)

In Kestra, we can now navigate to the Namespace we want to set up Vault as a secrets manager for and enter the configuration details:

![read-only-secrets-3](/docs/enterprise/read-only-secrets-3.png)

> After saving, we can move over to the **Secrets** tab and see which paths we have access to. Notice the lock icon indicating that read only is successfully turned on. No new secrets can be created from Kestra, and existing secrets are not editable.

Author: ```suggestion
After saving, we can move to the Secrets tab and see which paths we have access to. Notice the lock icon indicating that read-only is successfully turned on. No new secrets can be created from Kestra, and existing secrets are not editable.
```

![read-only-secrets-4](/docs/enterprise/read-only-secrets-4.png)

> In Vault, we know `my-app` is the secret that hosts the subkey we are looking for, in this case `NEON_PASSWORD`.

Author: ```suggestion
In Vault, we know `my-app` is the secret that hosts the subkey we are looking for, in this case, `NEON_PASSWORD`.
```
